server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java

```diff
@@ -0,0 +1,291 @@
+
+    action.createParam(PARAM_METRIC_KEY)
+      .setDescription("Metric key")
+      .setExampleValue("ncloc");
+
+    action.createParam(PARAM_VALUE)
+      .setRequired(true)
+      .setDescription(measureValueDescription())
+      .setExampleValue("47");
+
+    action.createParam(PARAM_DESCRIPTION)
+      .setDescription("Description")
+      .setExampleValue("Team size growing.");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    DbSession dbSession = dbClient.openSession(false);
+    String description = request.param(PARAM_DESCRIPTION);
+    long now = system.now();
+
+    try {
+      ComponentDto component = searchProject(dbSession, request);
+      MetricDto metric = searchMetric(dbSession, request);
+      checkPermissions(component);
+      checkMeasureDoesNotExistAlready(dbSession, component, metric);
+      CustomMeasureDto measure = new CustomMeasureDto()
+        .setComponentUuid(component.uuid())
+        .setComponentId(component.getId())
+        .setMetricId(metric.getId())
+        .setDescription(description)
+        .setCreatedAt(now);
+      setMeasureValue(measure, request, metric);
+      dbClient.customMeasureDao().insert(dbSession, measure);
+      dbSession.commit();
+
+      JsonWriter json = response.newJsonWriter();
+      writeMeasure(json, measure, component, metric, request.mandatoryParam(PARAM_VALUE));
+      json.close();
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+  }
+
+  private void checkPermissions(ComponentDto component) {
+    if (userSession.hasGlobalPermission(GlobalPermissions.SYSTEM_ADMIN)) {
+      return;
+    }
+
+    userSession.checkLoggedIn().checkProjectUuidPermission(UserRole.ADMIN, component.projectUuid());
+  }
+
+  private void checkMeasureDoesNotExistAlready(DbSession dbSession, ComponentDto component, MetricDto metric) {
+    int nbMeasuresOnSameMetricAndMeasure = dbClient.customMeasureDao().countByComponentIdAndMetricId(dbSession, component.uuid(), metric.getId());
+    if (nbMeasuresOnSameMetricAndMeasure > 0) {
+      throw new ServerException(HttpURLConnection.HTTP_CONFLICT, String.format("A measure already exists for project id '%s' and metric id '%d'", component.uuid(), metric.getId()));
+    }
+  }
+
+  private void writeMeasure(JsonWriter json, CustomMeasureDto measure, ComponentDto component, MetricDto metric, String measureWithoutInternalFormatting) {
+    json.beginObject();
+    json.prop(FIELD_ID, String.valueOf(measure.getId()));
+    json.prop(FIELD_METRIC_ID, String.valueOf(metric.getId()));
+    json.prop(FIELD_METRIC_KEY, metric.getKey());
+    json.prop(FIELD_PROJECT_ID, component.uuid());
+    json.prop(FIELD_PROJECT_KEY, component.key());
+    json.prop(FIELD_DESCRIPTION, measure.getDescription());
+    json.prop(FIELD_VALUE, measureWithoutInternalFormatting);
+    json.endObject();
+  }
+
+  private void setMeasureValue(CustomMeasureDto measure, Request request, MetricDto metric) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "setMeasureValue" is 16 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)